### PR TITLE
Refactor Note Selection

### PR DESCRIPTION
**Pull request by lselden**
*Edit: Accidentally added this PR onto bump of Avalonia version - added bump in separate [PR](https://github.com/stakira/OpenUtau/pull/452).*

This change introduces NoteSelectionViewModel, intended to make atomic changes to selection easier. It wraps up logic that will be used for keyboard-based navigation/range select. Relevant code points of existing mouse functionality also updated.

### CODE REVIEW

File: OpenUtau/ViewModels/NotesViewModel.cs [hunk 1]
```diff
@@ -91,7 +91,7 @@ public class NotesViewModel : ViewModelBase, ICmdSubscriber {
         private readonly ObservableAsPropertyHelper<double> smallChangeX;
         private readonly ObservableAsPropertyHelper<double> smallChangeY;
 
-        public readonly List<UNote> SelectedNotes = new List<UNote>();
+        public readonly NoteSelectionViewModel Selection = new NoteSelectionViewModel();
         private readonly HashSet<UNote> TempSelectedNotes = new HashSet<UNote>();
```
Comment (stakira) [inline OpenUtau/ViewModels/NotesViewModel.cs:95]: Shouldn't the new class also manage temporary selection?

Comment (lselden) [reply OpenUtau/ViewModels/NotesViewModel.cs]: Updated, now handles temp selection

File: OpenUtau/ViewModels/NoteSelectionViewModel.cs [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+        /// </summary>
+        public UNote? Head;
+        
+        public int Count => Notes.Count;
+        public bool IsEmpty => Notes.Count == 0;
+        private bool IsMultiple => Notes.Count > 1;
+        /// <summary>
+        /// true if the range is being resized from left side of selection rather than right
+        /// </summary>
+        private bool IsReversed => IsMultiple && Head == Notes.FirstOrDefault() ? true : false;
```
[the remark this answers is not in the document]
Comment (lselden) [reply OpenUtau/ViewModels/NoteSelectionViewModel.cs]: Agree - updated accordingly

**Comment by stakira** [comment]
Crashes in following cases:
- Paste notes then undo.
- Cut notes.
- Select notes then press del key.
- Delete note from context menu.

In general, modifying the collection inside foreach loop won't not work.

**Comment by lselden** [comment]
I was missing a `.ToList()` in there. Noted crash cases work now.

**Comment by stakira** [comment]
👍

**Pull request merged**